compiler/rustc_hir_typeck/src/writeback.rs

```diff
@@ -134,13 +134,32 @@ impl<'cx, 'tcx> WritebackCx<'cx, 'tcx> {
         self.fcx.tcx
     }
 
+    /// Checks if `ty` contains a `ty::Error` (i.e. `ty.references_error()`), but peering into
+    /// function signatures and ADTs.
+    fn deep_references_error(&self, ty: Ty<'tcx>) -> bool {
+        ty.references_error() || {
+            match ty.kind() {
+                ty::Adt(..) => ty.ty_adt_def().map(|adt| adt.references_error()).unwrap_or(false),
```

Comment:
This does cause lots of cycle errors, unfortunately. It seems necessary though, as soon as I change the test case so it isn't a tuple struct (and therefore doesn't have a ctor function), then you need this to avoid the ICE.